src/Appwrite/Resque/Worker.php

```diff
@@ -151,82 +152,100 @@ public function tearDown(): void
+        }
 
-        return $this->getDB(self::DATABASE_PROJECT, $projectId, $project->getInternalId());
+        return $this->getDB(self::DATABASE_PROJECT, $projectId, $project->getInternalId(), $project);
     }
 
     /**
      * Get console database
      * @return Database
+     * @throws Exception
      */
     protected function getConsoleDB(): Database
     {
         return $this->getDB(self::DATABASE_CONSOLE);
     }
 
     /**
-     * Get console database
-     * @param string $type One of (internal, external, console)
-     * @param string $projectId of internal or external DB
+     * Get database
+     * @param string $type One of (project, console)
+     * @param string $projectId of project or console DB
+     * @param string $projectInternalId
+     * @param Document|null $project
      * @return Database
+     * @throws Exception
      */
-    private function getDB(string $type, string $projectId = '', string $projectInternalId = ''): Database
-    {
+    private function getDB(
+        string $type,
+        string $projectId = '',
+        string $projectInternalId = '',
+        ?Document $project = null
+    ): Database {
         global $register;
 
-        $namespace = '';
         $sleep = DATABASE_RECONNECT_SLEEP; // overwritten when necessary
 
+        if ($project !== null) {
+            $projectId = $project->getId();
+            $projectInternalId = $project->getInternalId();
+        }
+
         switch ($type) {
             case self::DATABASE_PROJECT:
                 if (!$projectId) {
                     throw new \Exception('ProjectID not provided - cannot get database');
                 }
-                $namespace = "_{$projectInternalId}";
+                $namespace = "_$projectInternalId";
                 break;
             case self::DATABASE_CONSOLE:
                 $namespace = "_console";
                 $sleep = 5; // ConsoleDB needs extra sleep time to ensure tables are created
                 break;
             default:
                 throw new \Exception('Unknown database type: ' . $type);
-                break;
         }
 
         $attempts = 0;
 
-        do {
+        while (true) {
             try {
                 $attempts++;
                 $cache = new Cache(new RedisCache($register->get('cache')));
                 $database = new Database(new MariaDB($register->get('db')), $cache);
                 $database->setDefaultDatabase(App::getEnv('_APP_DB_SCHEMA', 'appwrite'));
                 $database->setNamespace($namespace); // Main DB
 
-                if (!empty($projectId) && !$database->getDocument('projects', $projectId)->isEmpty()) {
-                    throw new \Exception("Project does not exist: {$projectId}");
+                if (
+                    $project === null
+                    && !empty($projectId)
+                    && !$database->getDocument('projects', $projectId)->isEmpty()
```

Comment:
Will this condition ever be true ? 

If `project` is null, `projectId` will also be empty right? 
which means the condition `$project == null && !empty(projectId)` is always false ?

Reply:
You can pass `$projectId`, or the `$project` document, so this case will be true when a document was not passed, and a project with the given ID can not be found.

I have just added support for passing the entire document, to handle the race condition in this particular case where the project document is deleted from the database in the controller, but is required here in the worker